src/features/myAccounts/hooks/useAllSafes.ts

```diff
@@ -20,63 +21,71 @@ export type SafeItem = {
 
 export type SafeItems = SafeItem[]
 
-const useAddedSafes = () => {
-  const allAdded = useAppSelector(selectAllAddedSafes)
-  return allAdded
+export const _prepareAddresses = (
+  chainId: string,
+  allAdded: AddedSafesState,
+  allOwned: AllOwnedSafes,
+  allUndeployed: UndeployedSafesState,
+): string[] => {
+  const addedOnChain = Object.keys(allAdded[chainId] || {})
+  const ownedOnChain = allOwned[chainId] || []
+  const undeployedOnChain = Object.keys(allUndeployed[chainId] || {})
+
+  const combined = [...addedOnChain, ...ownedOnChain, ...undeployedOnChain]
+
+  return [...new Set(combined)]
 }
 
-export const useHasSafes = () => {
-  const { address = '' } = useWallet() || {}
-  const allAdded = useAddedSafes()
-  const hasAdded = !isEmpty(allAdded)
-  const [allOwned] = useAllOwnedSafes(!hasAdded ? address : '') // pass an empty string to not fetch owned safes
+export const _buildSafeItem = (
+  chainId: string,
+  address: string,
+  walletAddress: string,
+  allAdded: AddedSafesState,
+  allOwned: AllOwnedSafes,
+  allVisitedSafes: VisitedSafesState,
+  allSafeNames: AddressBookState,
+): SafeItem => {
+  const addedSafe = allAdded[chainId]?.[address]
+  const addedSafeOwners = addedSafe?.owners || []
+  const isPinned = Boolean(addedSafe) // Pinning a safe means adding it to the added safes storage
+
+  // Determine if the user is an owner
+  const isOwnerFromAdded = addedSafeOwners.some(({ value }) => sameAddress(walletAddress, value))
```

Comment:
This is just for CF safes, right?

Reply:
Good question. From what I can find is that we are using the `isReadOnly` flag for two things:

1. To show the Add network button
2. To show the queue actions

With that in mind I would expect the following behaviour:

CF Safes:
- Show the Add network button if the connected wallet is an owner regardless of pinned state 🔴 
- No queue actions required

Owned Safes:
- Show the Add network button regardless of pinned state ✅ 
- Show the queue actions regardless of pinned state ✅ 

Added but not owned safes:
- Show neither the Add network button nor the queue actions ✅ 

Anything I missed? I will implement the CF Safes case

Reply:
That behaviour makes sense, thanks for clarifying!

Reply:
>Good question. From what I can find is that we are using the isReadOnly flag for two things:
It's also used to display the read-only chip for the item in the safe list